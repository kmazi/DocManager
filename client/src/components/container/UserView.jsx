import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';

import { editUserDetail, changeInputValue } from '../../actions/userActions';
import UserView from '../presentation/UserView';

const formatDate = (datetime) => {
  if (datetime) {
    const date = datetime.substring(0, datetime.indexOf('T'));
    const formattedDate = new Date(date);
    return formattedDate.toDateString();
  }
};

const mapStateToProps = state => ({
  userName: state.authenticateUser.userName,
  updateStatus: state.authenticateUser.updateStatus,
  disabled: state.authenticateUser.disabled,
  userEmail: state.authenticateUser.updateEmail,
  userId: state.authenticateUser.userId,
  roleType: state.authenticateUser.roleType,
  createdAt: formatDate(state.authenticateUser.createdAt),
});

export default
connect(mapStateToProps, { editUserDetail,
  changeInputValue })(withRouter(UserView));
