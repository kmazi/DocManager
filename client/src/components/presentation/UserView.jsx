import React from 'react';
import PropTypes from 'prop-types';
import $ from 'jquery';
import Alert from 'sweetalert2';

const editDetail = (userId, editUser) => {
  const userDetail = {};
  if ($('#userEmail').val() !== '') {
    userDetail.email = $('#userEmail').val();
  }
  if ($('#oldpassword').val() !== '') {
    userDetail.oldPassword = $('#oldpassword').val();
  }
  if ($('#newpassword').val() !== '') {
    userDetail.password = $('#newpassword').val();
  }
  editUser(userDetail, userId).then((res) => {
    if (res === 'successful') {
      Alert({
        title: 'Update successful',
        text: 'You have successfully updated your profile',
        type: 'success',
        confirmButtonText: 'ok'
      });
    } else {
      Alert({
        title: 'Update error!',
        text: 'An error occurred while updating your profile',
        type: 'error',
        confirmButtonText: 'ok'
      });
    }
  });
};

const UserView = ({ userName, userId, userEmail, updateStatus,
  changeInputValue, createdAt, roleType, editUserDetail }) => (
    <div id="userview" className="container">
      <h5 className="center-align">SignUp Details</h5>
      <hr />
      <div className="row">
        <div className="col m2">
          <label htmlFor="username">Name:</label>
        </div>
        <div className="col m10">
          <input id="username" type="text" value={userName} disabled />
        </div>

      </div>
      <div className="row">
        <div className="col m2">
          <label htmlFor="userEmail">Email:</label>
        </div>
        <div className="col m10">
          <input
            id="userEmail"
            type="text"
            onChange={() => {
              changeInputValue($('#userEmail').val());
            }}
            value={userEmail}
            disabled
          />
        </div>

      </div>
      <div className="row">
        <div className="col m2">
          <label htmlFor="roleType">Department:</label>
        </div>
        <div className="col m10">
          <input id="roleType" type="text" value={roleType} disabled />
        </div>
      </div>

      <div id="passwordreset" className="hide">
        <div className="row">
          <div className="col m2">
            <label htmlFor="oldpassword">Old password:</label>
          </div>
          <div className="col m10">
            <input id="oldpassword" type="password" />
          </div>
        </div>
        <div className="row">
          <div className="col m2">
            <label htmlFor="newpassword">New password:</label>
          </div>
          <div className="col m10">
            <input id="newpassword" type="password" />
          </div>
        </div>
      </div>
      <div className="row">
        <div className="col m2">
          <label htmlFor="signUpDate">Registered on:</label>
        </div>
        <div className="col m10">
          <input id="signUpDate" type="text" value={createdAt} disabled />
        </div>
      </div>
      <div className="row">
        <button
          className="btn"
          id="editbtn"
          onClick={(event) => {
            const btnText = $('#editbtn').text() === 'Edit profile' ?
            'Back' : 'Edit profile';
            event.preventDefault();
            $('#userEmail').prop('disabled', (i, v) => (!v));
            $('#userEmail').focus();
            $('#passwordreset, #submitedit').toggleClass('hide');
            $('#editbtn').text(btnText);
          }}
        >Edit profile</button>
        <button
          id="submitedit"
          className="btn hide"
          onClick={(event) => {
            event.preventDefault();
            $('#userEmail').prop('disabled', (i, v) => (!v));
            $('#passwordreset, #submitedit').toggleClass('hide');
            $('#editbtn').text('Edit profile');
            editDetail(userId, editUserDetail);
          }}
        >{updateStatus}</button>
      </div>
    </div>
  );

UserView.propTypes = {
  userName: PropTypes.string.isRequired,
  userEmail: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  roleType: PropTypes.string.isRequired,
  userId: PropTypes.number.isRequired,
  updateStatus: PropTypes.string.isRequired,
  changeInputValue: PropTypes.func.isRequired,
  editUserDetail: PropTypes.func.isRequired,
};

export default UserView;
