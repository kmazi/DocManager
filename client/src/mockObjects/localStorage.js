/**
 * Mocks localstorage
 */
class LocalStorageMock {
  /**
   * the constructor
   */
  constructor() {
    this.store = {};
  }
/**
 * clears the localstorage
 * @return {void} returns null
 */
  clear() {
    this.store = {};
  }

/**
 * get item in the localstorage
 * @param {string} key - The key to the item in the localstorage
 * @return {string} returns the key to the item in localstorage
 */
  getItem(key) {
    return this.store[key];
  }
/**
 * set the value of an item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @param {string} value - The value of the item in the localstorage
 * @return {null} returns null
 */
  setItem(key, value) {
    this.store[key] = value;
  }

  /**
 * remove item in the localstorage class
 * @param {string} key - The key to the item in the localstorage
 * @return {void} returns nothing
 */
  removeItem(key) {
    delete this.store[key];
  }
}

global.localStorage = new LocalStorageMock();
export default global.localStorage;
