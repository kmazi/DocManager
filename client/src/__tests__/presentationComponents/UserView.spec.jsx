import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import UserView from '../../components/presentation/UserView';
import '../../mockObjects/localStorage';

describe('The UserView component:', () => {
  let response = 'successful';
  const props = {
    userName: 'jackson',
    userId: 2,
    userEmail: 'jackson@gmail.com',
    updateStatus: 'updating',
    changeInputValue: jest.fn(),
    createdAt: 'sun june 3, 2013',
    roleType: 'Learning',
    editUserDetail: jest.fn(() => Promise.resolve(response)),
  };
  test('should render the userview component correctly', () => {
    const component = renderer.create(<UserView {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should fire changeInputValue when userEmail input text changes',
  () => {
    const enzymeWrapper = mount(<UserView {...props} />);
    enzymeWrapper.find('#userEmail').props().onChange();
    expect(props.changeInputValue.mock.calls.length).toBe(1);
  });

  test('should fire editUserDetail when submitting the changes',
  () => {
    const event = {
      preventDefault: () => 'hello',
    };
    const enzymeWrapper = mount(<UserView {...props} />);
    enzymeWrapper.find('#submitedit').props().onClick(event);
    expect(props.editUserDetail.mock.calls.length).toBe(1);
  });

  test('should alert error message when editing function fails',
  () => {
    const event = {
      preventDefault: () => 'hello',
    };
    response = 'unsuccessful';
    const enzymeWrapper = mount(<UserView {...props} />);
    enzymeWrapper.find('#submitedit').props().onClick(event);
    const component = enzymeWrapper;
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });

  test('should unhide the submit button when the edit button is clicked',
  () => {
    const event = {
      preventDefault: () => 'hello',
    };
    const enzymeWrapper = mount(<UserView {...props} />);
    enzymeWrapper.find('#editbtn').props().onClick(event);
    expect(enzymeWrapper.find('#editbtn').hasClass('hide')).toBe(false);
  });
});
