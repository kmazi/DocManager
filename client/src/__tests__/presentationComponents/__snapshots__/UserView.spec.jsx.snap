// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The UserView component: should alert error message when editing function fails 1`] = `
<UserView
  changeInputValue={[Function]}
  createdAt="sun june 3, 2013"
  editUserDetail={[Function]}
  roleType="Learning"
  updateStatus="updating"
  userEmail="jackson@gmail.com"
  userId={2}
  userName="jackson"
>
  <div
    className="container"
    id="userview"
  >
    <h5
      className="center-align"
    >
      SignUp Details
    </h5>
    <hr />
    <div
      className="row"
    >
      <div
        className="col m2"
      >
        <label
          htmlFor="username"
        >
          Name:
        </label>
      </div>
      <div
        className="col m10"
      >
        <input
          disabled={true}
          id="username"
          type="text"
          value="jackson"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col m2"
      >
        <label
          htmlFor="userEmail"
        >
          Email:
        </label>
      </div>
      <div
        className="col m10"
      >
        <input
          disabled={true}
          id="userEmail"
          onChange={[Function]}
          type="text"
          value="jackson@gmail.com"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col m2"
      >
        <label
          htmlFor="roleType"
        >
          Department:
        </label>
      </div>
      <div
        className="col m10"
      >
        <input
          disabled={true}
          id="roleType"
          type="text"
          value="Learning"
        />
      </div>
    </div>
    <div
      className="hide"
      id="passwordreset"
    >
      <div
        className="row"
      >
        <div
          className="col m2"
        >
          <label
            htmlFor="oldpassword"
          >
            Old password:
          </label>
        </div>
        <div
          className="col m10"
        >
          <input
            id="oldpassword"
            type="password"
          />
        </div>
      </div>
      <div
        className="row"
      >
        <div
          className="col m2"
        >
          <label
            htmlFor="newpassword"
          >
            New password:
          </label>
        </div>
        <div
          className="col m10"
        >
          <input
            id="newpassword"
            type="password"
          />
        </div>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col m2"
      >
        <label
          htmlFor="signUpDate"
        >
          Registered on:
        </label>
      </div>
      <div
        className="col m10"
      >
        <input
          disabled={true}
          id="signUpDate"
          type="text"
          value="sun june 3, 2013"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <button
        className="btn"
        id="editbtn"
        onClick={[Function]}
      >
        Edit profile
      </button>
      <button
        className="btn hide"
        id="submitedit"
        onClick={[Function]}
      >
        updating
      </button>
    </div>
  </div>
</UserView>
`;

exports[`The UserView component: should render the userview component correctly 1`] = `
<div
  className="container"
  id="userview"
>
  <h5
    className="center-align"
  >
    SignUp Details
  </h5>
  <hr />
  <div
    className="row"
  >
    <div
      className="col m2"
    >
      <label
        htmlFor="username"
      >
        Name:
      </label>
    </div>
    <div
      className="col m10"
    >
      <input
        disabled={true}
        id="username"
        type="text"
        value="jackson"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col m2"
    >
      <label
        htmlFor="userEmail"
      >
        Email:
      </label>
    </div>
    <div
      className="col m10"
    >
      <input
        disabled={true}
        id="userEmail"
        onChange={[Function]}
        type="text"
        value="jackson@gmail.com"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col m2"
    >
      <label
        htmlFor="roleType"
      >
        Department:
      </label>
    </div>
    <div
      className="col m10"
    >
      <input
        disabled={true}
        id="roleType"
        type="text"
        value="Learning"
      />
    </div>
  </div>
  <div
    className="hide"
    id="passwordreset"
  >
    <div
      className="row"
    >
      <div
        className="col m2"
      >
        <label
          htmlFor="oldpassword"
        >
          Old password:
        </label>
      </div>
      <div
        className="col m10"
      >
        <input
          id="oldpassword"
          type="password"
        />
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col m2"
      >
        <label
          htmlFor="newpassword"
        >
          New password:
        </label>
      </div>
      <div
        className="col m10"
      >
        <input
          id="newpassword"
          type="password"
        />
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col m2"
    >
      <label
        htmlFor="signUpDate"
      >
        Registered on:
      </label>
    </div>
    <div
      className="col m10"
    >
      <input
        disabled={true}
        id="signUpDate"
        type="text"
        value="sun june 3, 2013"
      />
    </div>
  </div>
  <div
    className="row"
  >
    <button
      className="btn"
      id="editbtn"
      onClick={[Function]}
    >
      Edit profile
    </button>
    <button
      className="btn hide"
      id="submitedit"
      onClick={[Function]}
    >
      updating
    </button>
  </div>
</div>
`;
